### ISSUE TITLE
Non-deterministic memory issues

### ISSUE THREAD

**Issue by RReverser** [colmnet/ragel#10]
Periodically getting this when running Ragel, doesn't matter whether for compiling or generating graph, and doesn't seem to depend on [size of] input files:

```
ragel(27505,0x7fff798bb000) malloc: *** error for object 0x7f9a02503270: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

```
ragel(27514,0x7fff798bb000) malloc: *** error for object 0x7f8bb0e027b0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

```
ragel(27523,0x7fff798bb000) malloc: *** error for object 0x7ff13b400030: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

Usually re-running helps, but might be a sign of some serious error.

**Comment by RReverser** [comment]
Tried to debug with LLDB, looks like exception occurs here:

```cpp
InputData::~InputData()
{
	includeDict.empty();
	inputItems.empty();
	parseDataList.empty();
	sectionList.empty();

	for ( Vector<const char**>::Iter fns = streamFileNames; fns.lte(); fns++ ) {
		const char **ptr = *fns;
		while ( *ptr != 0 ) {
			::free( (void*)*ptr ); // <---------------------
			ptr += 1;
		}
		free( (void*) *fns );
	}
```

**Comment by RReverser** [comment]
Some more context from LLDB:

```
* thread #1: tid = 0x1946d0, 0x000000010004ea58 ragel`InputData::~InputData(this=0x00007fff5fbff4c8) + 216 at inputdata.cc:54, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x000000010004ea58 ragel`InputData::~InputData(this=0x00007fff5fbff4c8) + 216 at inputdata.cc:54
   51  		for ( Vector<const char**>::Iter fns = streamFileNames; fns.lte(); fns++ ) {
   52  			const char **ptr = *fns;
   53  			while ( *ptr != 0 ) {
-> 54  				::free( (void*)*ptr );
   55  				ptr += 1;
   56  			}
   57  			free( (void*) *fns );
(lldb) frame variable *ptr
(const char *) *ptr = 0x00000001006001b0 ""
(lldb) continue
Process 83441 resuming
ragel(83441,0x7fff7472a000) malloc: *** error for object 0x1006001b0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Process 83441 stopped
* thread #1: tid = 0x1946d0, 0x00007fff856eef3e libsystem_malloc.dylib`malloc_error_break, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
    frame #0: 0x00007fff856eef3e libsystem_malloc.dylib`malloc_error_break
libsystem_malloc.dylib`malloc_error_break:
->  0x7fff856eef3e <+0>: pushq  %rbp
    0x7fff856eef3f <+1>: movq   %rsp, %rbp
    0x7fff856eef42 <+4>: nop
    0x7fff856eef43 <+5>: nopl   (%rax)
```

Appears that iterator contains some empty string which cannot be released (maybe constant?).

**Comment by RReverser** [comment]
@colmnet I'm stuck with debugging it further as not very familiar with codebase. Can't figure out under which conditions it happens or why re-running several times helps, but this is weird :/

**Comment by colmnet** [comment]
Have you gotten this problem lately? There was one recent fix which may handle the problem.

Background: It's to do with memory management of file names referenced by parsing locations. I am using colm as a parser generator. It automatically manages parsing locations for error reporting, and this includes file name strings. The problem is that we need parsing locations to survive long past the parsing context for error reporting and propagation to the compiled output. Initially I just leaked the file name strings. Enter some cleanup and we now release them to the caller, which can then free them when finished. Unfortunately I got used to leaking them so over time some bugs crept into the file name allocation. As a short-term fix we can just leak them again by refraining from delete.

**Comment by colmnet** [comment]
If you've got an input that fails we can verify (even if it fails only occasionally).

**Comment by RReverser** [comment]
Yes, I actually started getting this couple of weeks ago and didn't stop yet (always on github master so far). Can't share our state machine publicly I'm afraid :(

**Comment by colmnet** [comment]
Can you tell me if it has 1) includes 2) exports? If it has includes, is it a machine in the same file, or another file?

**Comment by RReverser** [comment]
Yes, definitely using bunch of includes to split machine across couple of files.

**Comment by RReverser** [comment]
Not sure if this helps, but I'm getting it (occasionally) even during the make of Ragel itself:
```
/usr/local/bin/ragel -G2 -I. -o rlscan.cc rlscan.rl
ragel(9439,0x7fff79c4d000) malloc: *** error for object 0x7fadb0d02f60: pointer being freed was not allocated
```

**Comment by RReverser** [comment]
Interesting side effect that seems to be related: during the runtests (and very rarely manually) I see that some `#line` directives are generated with a garbage instead of filename so GCC fails to parse the following content. Example from `call1.c`:

```c

#line 1 "working/call1.rl"
/*
* @LANG: c
* @PROHIBIT_FEATFLAGS: --var-backend
*/

#include <stdio.h>
#include <string.h>

int num = 0;

struct test
{
	int cs, top, stack[32];
};


#line 40 "working/call1.rl"



static const char _test_actions[] = {
	0, 1, 0, 1, 1, 1, 2, 1,
	3, 0
};

static const char _test_trans_keys[] = {
	1, 0, 3, 6, 2, 4, 0, 0,
	5, 5, 2, 2, 8, 11, 12, 12,
	12, 12, 11, 11, 12, 12, 4, 4,
	9, 9, 8, 8, 13, 13, 8, 8,
	10, 10, 1, 0, 11, 11, 7, 7,
	7, 7, 1, 0, 1, 0, 1, 0,
	1, 0, 1, 0, 0
};

static const char _test_indicies[] = {
	0, 0, 1, 2, 3, 0, 0, 4,
	5, 6, 7, 1, 1, 8, 9, 10,
	11, 12, 13, 13, 14, 15, 16, 17,
	19, 20, 21, 0
};

static const char _test_char_class[] = {
	0, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 1, 1, 1, 1, 2, 1,
	1, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 1, 1, 1, 1, 3, 4,
	4, 3, 3, 3, 3, 3, 3, 3,
	1, 1, 1, 1, 1, 1, 1, 5,
	1, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 1, 1, 1, 6, 1, 1,
	1, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 7, 8, 1, 1, 1, 1,
	1, 9, 1, 1, 10, 11, 1, 1,
	12, 1, 1, 1, 13, 0
};

static const char _test_index_offsets[] = {
	0, 0, 4, 7, 8, 9, 10, 14,
	15, 16, 17, 18, 19, 20, 21, 22,
	23, 24, 24, 25, 26, 27, 27, 27,
	27, 27, 0
};

static const char _test_index_defaults[] = {
	0, 1, 1, 1, 1, 1, 1, 1,
	1, 1, 1, 1, 1, 1, 1, 1,
	1, 18, 1, 1, 1, 22, 1, 1,
	1, 1, 0
};

static const char _test_cond_targs[] = {
	2, 0, 4, 3, 23, 5, 6, 7,
	12, 8, 9, 10, 11, 3, 14, 15,
	16, 17, 24, 19, 20, 21, 25, 0
};

static const char _test_cond_actions[] = {
	7, 0, 0, 1, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 3, 0, 0, 0, 5, 0
};

static const int test_start = 1;
static const int test_first_final = 23;
static const int test_error = 0;

static const int test_en_even = 13;
static const int test_en_odd = 18;
static const int test_en_fail = 22;
static const int test_en_main = 1;


#line 43 "@�
�"


void test_init( struct test *fsm )
{
	num = 0;
	
	{
		fsm->cs = (int)test_start;
		fsm->top = 0;
	}
	
	#line 48 "@�
	�"
	
}

void test_execute( struct test *fsm, const char *data, int len )
{
	const char *p = data;
	const char *pe = data+len;
	
	
	{
		int _trans = 0;
		const char *_acts;
		unsigned int _nacts;
		const char *_keys;
		const char *_inds;
		{
			
			if ( p == pe )
			goto _test_eof;
			if ( fsm->cs == 0 )
			goto _out;
			_resume:  {
				_keys = ( _test_trans_keys + ((fsm->cs<<1)));
				_inds = ( _test_indicies + (_test_index_offsets[fsm->cs]));
				
				if ( ( (*( p))) <= 118 && ( (*( p))) >= 10 )
				{
					int _ic = (int)_test_char_class[( (*( p))) - 10];
					if ( _ic <= (int)(*( _keys+1)) && _ic >= (int)(*( _keys)) )
					_trans = (int)(*( _inds + (int)( _ic - (int)(*( _keys)) ) )); 
					else
					_trans = (int)_test_index_defaults[fsm->cs];
				}
				else {
					_trans = (int)_test_index_defaults[fsm->cs];
				}
				
				goto _match_cond;
			} _match_cond:  {
				fsm->cs = (int)_test_cond_targs[_trans];
				
				if ( _test_cond_actions[_trans] == 0 )
				goto _again;
				
				_acts = ( _test_actions + (_test_cond_actions[_trans]));
				_nacts = (unsigned int)(*( _acts));
				_acts += 1;
				while ( _nacts > 0 ) {
					switch ( (*( _acts)) )
					{
						case 0:  {
							{
								#line 20 "working/call1.rl"
								
								if ( num & 1 )
								{fsm->stack[fsm->top] = fsm->cs; fsm->top += 1;fsm->cs = (18); goto _again;}
								else
								{fsm->stack[fsm->top] = fsm->cs; fsm->top += 1;fsm->cs = 13; goto _again;}}
							break; }
						case 1:  {
							{
								#line 28 "working/call1.rl"
								{p = p - 1; } {fsm->top -= 1;fsm->cs = fsm->stack[fsm->top];goto _again;}}
							break; }
						case 2:  {
							{
								#line 29 "working/call1.rl"
								{p = p - 1; } {fsm->top -= 1;fsm->cs = fsm->stack[fsm->top];goto _again;}}
							break; }
						case 3:  {
							{
								#line 30 "working/call1.rl"
								num = num * 10 + ((( (*( p)))) - '0'); }
							break; }
					}
					_nacts -= 1;
					_acts += 1;
				}
				
				
			} _again:  {
				if ( fsm->cs == 0 )
				goto _out;
				p += 1;
				if ( p != pe )
				goto _resume;
			} _test_eof:  { {}
			} _out:  { {}
			}}
	}
	
	#line 56 "working/call1.rl"
	
}

int test_finish( struct test *fsm )
{
	if ( fsm->cs == test_error )
	return -1;
	if ( fsm->cs >= test_first_final )
	return 1;
	return 0;
}

#define BUFSIZE 1024

void test( char *buf )
{   
	struct test test;
	test_init( &test );
	test_execute( &test, buf, strlen(buf) );
	if ( test_finish( &test ) > 0 )
	printf( "ACCEPT\n" );
	else
	printf( "FAIL\n" );
}

int main()
{
	test( "78 even\n" );
	test( "89 odd\n" );
	test( "1 even\n" );
	test( "0 odd\n" );
	test( "OA ok\n" );
	test( "OA error1\n" );
	test( "OA error2\n" );
	
	return 0;
}
```

and corresponding errors:

```
gcc -Wall -O3 -I. -Wno-variadic-macros -o working/call1.bin working/call1.c
working/call1.rl:118:10: warning: missing terminating '"' character [-Winvalid-pp-token]
#line 43 "@<U+001C><F0>
         ^
working/call1.rl:118:10: error: invalid filename for #line directive
working/call1.rl:119:1: error: source file is not valid UTF-8
<A5><U+007F>"
^
working/call1.rl:119:2: error: expected identifier or '('
<A5><U+007F>"
    ^
working/call1.rl:119:3: warning: missing terminating '"' character [-Winvalid-pp-token]
<A5><U+007F>"
            ^
working/call1.rl:131:11: warning: missing terminating '"' character [-Winvalid-pp-token]
        #line 48 "@<U+001C><F0>
                 ^
working/call1.rl:131:11: error: invalid filename for #line directive
working/call1.rl:132:2: error: source file is not valid UTF-8
        <A5><U+007F>"
        ^
working/call1.rl:132:4: warning: missing terminating '"' character [-Winvalid-pp-token]
        <A5><U+007F>"
                    ^
working/call1.rl:73:2: warning: implicit declaration of function 'test_init' is invalid in C99 [-Wimplicit-function-declaration]
        test_init( &test );
```

**Comment by colmnet** [comment]
Cannot reproduce this failure exactly (heap corruption manifests differently for me), but I expect this to be fixed with 9be749f2fb04932c0c6d49e81651bcf0c6be5fec. Please open a new issue if not.
